mflowgen/tile_array/custom-init/outputs/floorplan.tcl

```diff
@@ -0,0 +1,160 @@
+  set tiles($row,$col,height) [dbGet [dbGet -p top.insts.name $tile_name -i 0].cell.size_y]
+
+  # grid height = max row value
+  if {$row > $max_row} {
+    set max_row $row
+  }
+  if {$row < $min_row} {
+    set min_row $row
+  }
+  if {$col > $max_col} {
+    set max_col $col
+  }
+  if {$col < $min_col} {
+    set min_col $col
+  }
+}
+
+# Get grid height/width from max_row/col
+set grid_num_rows [expr $max_row - $min_row + 1]
+set grid_num_cols [expr $max_col - $min_col + 1]
+# Multiply separation params by respective pitches to calculate actual separation numbers
+set tile_separation_x [expr $tile_separation_x * $horiz_pitch]
+set tile_separation_y [expr $tile_separation_y * $vert_pitch]
+# Calculate the height and width of the full array of tiles including any inter-tile spacing
+set grid_height 0
+for {set i $min_row} {$i <= $max_row} {incr i} {
+  set grid_height [expr $grid_height + $tiles($i,$min_col,height) + $tile_separation_y]
+}
+# Subtract the spacing for the last tile for correct size
+set grid_height [expr $grid_height - $tile_separation_y]
+
+set grid_width 0
+for {set i $min_col} {$i <= $max_col} {incr i} {
+  set grid_width [expr $grid_width + $tiles($min_row,$i,width) + $tile_separation_x]
+}
+# Subtract the spacing for the last tile for correct size
+set grid_width [expr $grid_width - $tile_separation_x]
+
+# Now use the width of the grid and the specified margins to calculate floorplan size
+set grid_margin_t [expr $grid_margin_t * $vert_pitch]
+set grid_margin_b [expr $grid_margin_b * $vert_pitch]
+set grid_margin_l [expr $grid_margin_l * $horiz_pitch]
+set grid_margin_r [expr $grid_margin_r * $horiz_pitch]
+
+floorPlan -s [expr $grid_width + $grid_margin_l + $grid_margin_r] \
+             [expr $grid_height + $grid_margin_t + $grid_margin_b] \
+             $core_margin_l $core_margin_b $core_margin_r $core_margin_t
+
+setFlipping s
+
+# Now, actually place the CGRA tiles
+set start_x [expr $core_margin_l + $grid_margin_l]
+set start_y [expr $core_margin_b + $grid_margin_b]
+set y_loc $start_y
+for {set row $max_row} {$row >= $min_row} {incr row -1} {
+  set x_loc $start_x
+  for {set col $min_col} {$col <= $max_col} {incr col} {
+    set tiles($row,$col,x_loc) $x_loc
+    set tiles($row,$col,y_loc) $y_loc
+    placeInstance $tiles($row,$col,name) $x_loc $y_loc -fixed
+    set x_loc [expr $x_loc + $tiles($row,$col,width) + $tile_separation_x]
+  }
+  set y_loc [expr $y_loc + $tiles($row,$min_col,height) + $tile_separation_y]
+}
+
+addHaloToBlock -allMacro [expr $horiz_pitch * 3] $vert_pitch [expr $horiz_pitch * 3] $vert_pitch
+
+# Manually connect all of the tile_id pins
+selectPin *tile_id*
+set tile_id_pin [dbGet selected -i 0]
+set pin_depth [dbGet $tile_id_pin.cellTerm.pins.allShapes.shapes.rect_sizex -i 0]
+set connection_layer [dbGet $tile_id_pin.layer.name]
+# Make the connection between tie and tile id pins 2x the min width for that layer
+set connection_width [expr 2 * [dbGet $tile_id_pin.layer.minWidth]]
```

Comment:
Since you left 5*horiz_pitch routing gap in the tile flow, I guess there is enough space for this? Does minWidth == horiz_pitch? Would be good to understand the connection between the routing blockage size in the tile flow and the spacings used here..

Reply:
These things are not the same. horiz_pitch, as we've been using it, refers to the unit width of the standard cells. Min width here, is the minimum allowed width of any wire on that metal layer. The metal layer also has an associated pitch, which I suppose I could use here instead of min_width.

Reply:
I guess my issue is that it's not clear that you've created enough routing blockage in the tile flow to have room for this minWidth-sized wire ... it seems like you've just created a routing blockage around the pin area .. maybe because the wire is the minimum width there is a guarantee no other wire would overlap it because otherwise it would need to pass through the blockage? Would be good to document your reasoning here. And even if there was no overlap, couldn't some other wire in the tile be directly adjacent to this one, possibly causing DRCs? It seems like you've only added extra space on the other side of the tile.

Reply:
I can't add extra space on the pinning side, because then the router will not be able to connect the tile_id inputs to their internal destinations. The wire I add for the manual connection starts at the very edge fo the tile and goes in 2*minWidth, which, for our current pinning is enough less than pinDepth not to cause issues. Using minWidth wires to make the connections caused a geometry-related DRC error, which is why I went with 2* minWIdth.

Reply:
Awesome, yeah if you can just add a comment that 2xminWidth is less than the pinDepth I think it will make a lot more sense. Otherwise everything else looks good.